pytorch_lightning/trainer/training_tricks.py

```diff
@@ -80,3 +86,175 @@ def configure_accumulated_gradients(self, accumulate_grad_batches):
+                        new_size = _adjust_batch_size(self, factor=0.5, string='failed')
+                    break
+                else:
+                    raise  # some other error not memory related
+
+        # If in binsearch mode, further refine the search for optimal batch size
+        if mode == 'binsearch':
+            while True:
+                garbage_collection_cuda()
+                self.global_step = 0  # reset after each try
+                try:
+                    # Try fit
+                    self.fit(model)
+                    count += 1
+                    if count > n_max_try:
+                        break
+
+                    # Adjust batch size
+                    low = new_size
+                    midval = (high + low) // 2
+                    new_size = _adjust_batch_size(self, value=midval, string='succeeded')
+                except RuntimeError as exception:
+                    # Only these errors should trigger an adjustment
+                    if is_OOM_error(exception):
+                        garbage_collection_cuda()
+                        high = new_size
+                        if high - low <= 1:
+                            break
+                        midval = (high + low) // 2
+                        new_size = _adjust_batch_size(self, value=midval, string='failed')
+                    else:
+                        raise  # some other error not memory related
+
+        garbage_collection_cuda()
+        log.info(f'Finished batch size finder, will continue with full run using batch size {new_size}')
+
+        # Restore initial state of model
+        self.restore(str(save_path), on_gpu=self.on_gpu)
+        os.remove(save_path)
+
+        # Finish by resetting variables so trainer is ready to fit model
+        self.max_steps = max_steps
+        self.weights_summary = weights_summary
+        self.logger = logger
+        self.callbacks = callbacks
+        self.checkpoint_callback = checkpoint_callback
+
+        return new_size
+
+
+def _adjust_batch_size(trainer,
+                       factor: float = 1.0,
+                       value: Optional[int] = None,
+                       string: str = None):
+    """ Function for adjusting the batch size. It is expected that the user
+        has provided a model that has a hparam field called `batch_size` i.e.
+        `model.hparams.batch_size` should exist.
+
+    Args:
+        trainer: instance of pytorch_lightning.Trainer
+
+        factor: value which the old batch size is multiplied by to get the
+            new batch size
+
+        value: if a value is given, will override the batch size with this value.
+            Note that the value of `factor` will not have an effect in this case
+
+        string: either `succeeded` or `failed`. Used purely for logging
+
+    """
+    trainer_arg = 'batch_size'
+
+    model = trainer.get_model()
+
+    if hasattr(model.hparams, trainer_arg):
```

Comment:
from a user perspective, it is strange that you need to have it set even you asked to find the best/largest...

Reply:
@PyTorchLightning/core-contributors ^^

Reply:
@Borda I agree it is a bit strange. However, your question made me realize that it is necessary to do it this way, because we need the users `train_dataloader` to explicit depend on a value that we can alter before it is constructed, since you cannot change the `batch_size` of a pytorch dataloader after it is constructed. For this reason, this feature does not work when the user pass in a dataloader to `.fit()` (because it is already constructed), only when it is defined in the `train_dataloader` method.